### PR TITLE
musescore: re-add USE_SYSTEM_FREETYPE functionality

### LINKED ISSUE #244409: Musescore 4.1.0 segfault on launch
### Describe the bug
Running musescore 4.1.0 from nixpkgs master causes a segfault on launch.

### Steps To Reproduce
1. Run musescore via `nix run nixpkgs/master#musescore`
2. Application starts to launch, with a small window appearing for a brief second, before crashing.

### Log file
```
nix run nixpkgs/master#musescore
13:51:44.623 | INFO  | main_thread | GlobalModule | onPreInit: log path: /home/huantian/.local/share/MuseScore/MuseScore4/logs/MuseScore_230719_135144.log
13:51:44.623 | INFO  | main_thread | GlobalModule | onPreInit: === Started MuseScore 4.1.0, build number  ===
13:51:44.624 | WARN  | main_thread | IpcSocket  | connect: failed connect to server
13:51:44.624 | WARN  | main_thread | IpcSocket  | connect: failed connect to server
13:51:44.624 | INFO  | main_thread | IpcSocket  | connect: success connected to ipc server
13:51:44.624 | INFO  | 139704377599680 | IpcServer  | listen: id: "312d7b4729a348df8a494861a9a2e50a" 
13:51:44.639 | WARN  | main_thread | DiagnosticsModule | onInit: crash handling disabled
13:51:44.657 | ERROR | main_thread | MuseSamplerLibHandler | MuseSamplerLibHandler: Unable to open MuseSampler library, path: /home/huantian/.local/share/MuseSampler/lib/libMuseSamplerCoreLib.so
13:51:44.657 | ERROR | main_thread | MuseSamplerResolver | checkLibrary: Incompatible MuseSampler library; ignoring
13:51:44.657 | ERROR | main_thread | MuseSamplerLibHandler | MuseSamplerLibHandler: Unable to open MuseSampler library, path: libMuseSamplerCoreLib.so
13:51:44.657 | ERROR | main_thread | MuseSamplerResolver | checkLibrary: Incompatible MuseSampler library; ignoring
13:51:44.657 | WARN  | main_thread | Qt         | QIODevice::read (QFile, "/home/huantian/.local/share/MuseScore/MuseScore4/shortcuts.xml"): device not open
13:51:44.658 | WARN  | main_thread | Qt         | QIODevice::read (QFile, "/home/huantian/.local/share/MuseScore/MuseScore4/midi_mappings.xml"): device not open
13:51:44.666 | WARN  | main_thread | AbstractCloudService | readTokens: Could not find the tokens file: /home/huantian/.local/share/MuseScore/MuseScore4/musescorecom_cred.dat
13:51:44.666 | WARN  | main_thread | AbstractCloudService | readTokens: Could not find the tokens file: /home/huantian/.local/share/MuseScore/MuseScore4/audiocom_cred.dat
13:51:44.844 | ERROR | main_thread | PluginsConfiguration | updatePluginsConfiguration: [401] The file does not exist
[1]    50556 segmentation fault (core dumped)  nix run nixpkgs/master#musescore
```

### Notify maintainers
@doronbehar @turion @vandenoever 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.4.3, NixOS, 23.11 (Tapir), 23.11.20230718.684c17c`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.15.1`
 - channels(root): `""`
 - channels(huantian): `""`
 - nixpkgs: `/etc/nixpkgs/channels/nixpkgs`

```

### PR DESCRIPTION
**Pull request by huantianad**
## Description of changes
Fixes #244409.

Upstream removed the option to use system freetype library in v4.1.0, causing the app to crash on systems when the outdated bundled freetype tries to load the Noto Sans font. For more info on the crash itself, see #244409 and https://github.com/musescore/MuseScore/issues/18795.

For now, re-add the option ourselves. The fix has been merged upstream, so we can remove this patch with the next version. In the future, we may replace the other bundled thirdparty libs with system libs, see https://github.com/musescore/MuseScore/issues/11572.

@doronbehar @turion 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/musescore/default.nix [hunk 1]
```diff
@@ -64,6 +64,14 @@ in stdenv'.mkDerivation rec {
       url = "https://github.com/doronbehar/MuseScore/commit/f48448a3ede46f5a7ef470940072fbfb6742487c.patch";
       hash = "sha256-UEc7auscnW0KMfWkLKQtm+UstuTNsuFeoNJYIidIlwM=";
     })
+    # Upstream removed the option to use system freetype in v4.1.0,
+    # which caused crashes as the bundled version is outdated, see #244409.
+    # For now apply the patch manually. The fix has been merged upstream,
+    # so we can remove this patch with the next version.
```
Comment (doronbehar) [inline pkgs/applications/audio/musescore/default.nix:70]: Links that I think are important to put here:

- https://github.com/musescore/MuseScore/issues/11572
- https://github.com/musescore/MuseScore/issues/18795

**Comment by huantianad** [comment]
Did basic testing locally, built and ran the package, as well as ran the NixOS tests (`nix build .#nixosTests.musescore`). Cannot test for darwin or aarch64, but ofborg seem(ed) to be happy.

**Comment by doronbehar** [comment]
> Edit: I also noticed that we aren't running MuseScore's cmake tests, could it be possible to get those running at some point?

Simply setting `doCheck = true;` is worth trying.

**Comment by huantianad** [comment]
Hm all the tests seem to fail with the same error:
```
musescore> running tests
musescore> check flags: -j16 test
musescore> [0/1] Running tests...
musescore> Test project /build/source/build
musescore>       Start  1: global_tests
musescore>       Start  2: accessibility_tests
musescore>       Start  3: audio_test
musescore>       Start  4: draw_tests
musescore>       Start  5: mpe_test
musescore>       Start  6: ui_tests
musescore>       Start  7: braille_tests
musescore>       Start  8: diagnostics_tests
musescore>       Start  9: engraving_tests
musescore>       Start 10: iex_bb_tests
musescore>       Start 11: iex_bww_tests
musescore>       Start 12: iex_capella_tests
musescore>       Start 13: iex_midi_tests
musescore>       Start 14: iex_musicxml_tests
musescore>       Start 15: iex_guitarpro_tests
musescore>       Start 16: plugins_tests
musescore>  1/17 Test #15: iex_guitarpro_tests ..............Subprocess aborted***Exception:   0.11 sec
musescore> qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
musescore> This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
musescore> Available platform plugins are: wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx.
musescore>       Start 17: project_test
musescore>  2/17 Test  #3: audio_test .......................Subprocess aborted***Exception:   0.12 sec
musescore> qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
musescore> This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
musescore> Available platform plugins are: wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx.

...
musescore> 16/17 Test  #9: engraving_tests ..................Subprocess aborted***Exception:   0.13 sec
musescore> qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
musescore> This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
musescore> Available platform plugins are: wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx.
musescore> 17/17 Test #17: project_test .....................Subprocess aborted***Exception:   0.22 sec
musescore> qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
musescore> This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
musescore> Available platform plugins are: wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx.
musescore> 0% tests passed, 17 tests failed out of 17
musescore> Total Test time (real) =   0.35 sec
musescore> The following tests FAILED:
musescore>        1 - global_tests (Subprocess aborted)
musescore>        2 - accessibility_tests (Subprocess aborted)
musescore>        3 - audio_test (Subprocess aborted)
musescore>        4 - draw_tests (Subprocess aborted)
musescore>        5 - mpe_test (Subprocess aborted)
musescore>        6 - ui_tests (Subprocess aborted)
musescore>        7 - braille_tests (Subprocess aborted)
musescore>        8 - diagnostics_tests (Subprocess aborted)
musescore>        9 - engraving_tests (Subprocess aborted)
musescore>       10 - iex_bb_tests (Subprocess aborted)
musescore>       11 - iex_bww_tests (Subprocess aborted)
musescore>       12 - iex_capella_tests (Subprocess aborted)
musescore>       13 - iex_midi_tests (Subprocess aborted)
musescore>       14 - iex_musicxml_tests (Subprocess aborted)
musescore>       15 - iex_guitarpro_tests (Subprocess aborted)
musescore>       16 - plugins_tests (Subprocess aborted)
musescore>       17 - project_test (Subprocess aborted)
musescore> Errors while running CTest
musescore> FAILED: CMakeFiles/test.util
musescore> cd /build/source/build && /nix/store/p6nyqsw5qbg114gprcj4ci8j4nx9nhnb-cmake-3.26.4/bin/ctest --force-new-ctest-process
musescore> ninja: build stopped: subcommand failed.
```

Not sure if these are all the tests there are to run, given https://github.com/musescore/MuseScore/blob/efd683db329e2588a04d1828a2c3b140d0a25a55/CMakeLists.txt#L48-L49

**Comment by doronbehar** [comment]
> Not sure if these are all the tests there are to run, given [musescore/MuseScore@`efd683d`/CMakeLists.txt#L48-L49](https://github.com/musescore/MuseScore/blob/efd683db329e2588a04d1828a2c3b140d0a25a55/CMakeLists.txt#L48-L49)

See however: https://github.com/musescore/MuseScore/blob/efd683db329e2588a04d1828a2c3b140d0a25a55/CMakeLists.txt#L220

The failure you observed might be fixable if you'd do something like this:

```
diff --git i/pkgs/applications/audio/musescore/default.nix w/pkgs/applications/audio/musescore/default.nix
index ecc2ccf875f2..034f1411b440 100644
--- i/pkgs/applications/audio/musescore/default.nix
+++ w/pkgs/applications/audio/musescore/default.nix
@@ -131,6 +131,11 @@ in stdenv'.mkDerivation rec {
     alsa-lib
   ];
 
+  # Tests link to qt and require this
+  preCheck = ''
+    export QT_PLUGIN_PATH=${qtbase.bin}/${qtbase.qtPluginPrefix}
+  '';
+
   postInstall = ''
     # Remove unneeded bundled libraries and headers
     rm -r $out/{include,lib}
```

Haven't tested.

**Comment by huantianad** [comment]
Hm that doesn't seem to work, it looks like the tests attempt to actually display a MuseScore window, which requires it to connect to the X server, which it cannot do. Running with `QT_DEBUG_PLUGINS=1` shows this:

```
...
musescore> Got keys from plugin meta data ("xcb")
musescore> QFactoryLoader::QFactoryLoader() checking directory path "/build/source/build/src/project/tests/platforms" ...
musescore> loaded library "/nix/store/kqghllpfj5kcfyns1aldl7lmcszvz8nz-qtbase-5.15.9-bin/lib/qt-5.15.9/plugins/platforms/libqxcb.so"
musescore> qt.qpa.xcb: could not connect to display
musescore> qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
musescore> This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
musescore> Available platform plugins are: wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, xcb.
```

**Comment by doronbehar** [comment]
I see. Thanks for investigating that! Let's add a `doCheck = false;` and write a comment that tests require a running x window system, and that we have a nixosTest instead.

**Comment by turion** [comment]
I could successfully build and run. It didn't crash like before.

**Pull request merged**